### PR TITLE
Fix dialogs pushed to `DialogOverlay` too early potentially doing cross-thread transforms

### PR DESCRIPTION
**Pull request by peppy**
As seen at https://github.com/ppy/osu/runs/6287700705?check_suite_focus=true.

### CODE REVIEW

File: osu.Game/Overlays/DialogOverlay.cs [hunk 1]
```diff
@@ -88,16 +94,14 @@ protected override void PopIn()
         protected override void PopOut()
         {
             base.PopOut();
+            this.FadeOut(PopupDialog.EXIT_DURATION, Easing.InSine);
 
-            lowPassFilter.CutoffTo(AudioFilter.MAX_LOWPASS_CUTOFF, 100, Easing.InCubic);
+            // PopOut is called as part of VisibilityContainer's initialisation logic, but we don't want it to interact with a potentially waiting dialog.
+            if (!IsLoaded) return;
```
Comment (frenzibyte) [inline osu.Game/Overlays/DialogOverlay.cs:100]: This change should be irrelevant now that the push is delayed until the dialog is loaded, yeah?

**Review by frenzibyte — COMMENTED** [review]
I'm still not sure how this fixes the test failure at hand, given that the `DialogOverlay` loads the popup dialog at its current thread synchronously, and also loads `lowPassFilter` at its current thread synchronously as well...

In addition, the test failures are relevant :|

**Comment by peppy** [comment]
I've pushed another solution but the reverts `CurrentDialog` to immediately be in the correct state, as used to be the case. It seems that this is required to not only fix the failing test, but handle the scenario where `DialogOverlay` may not be finished async loading, but a check-for-dialog pattern is used, ie.

https://github.com/ppy/osu/blob/832d37b2c2c9b5f2cde6bec90804e6dab36e8e0e/osu.Game/Screens/Edit/Editor.cs#L600-L601

Unfortunately this makes the test still pass without the fix, but not much can be done about that..

File: osu.Game/Overlays/DialogOverlay.cs [hunk 2]
```diff
@@ -49,18 +49,24 @@ public void Push(PopupDialog dialog)
...
             CurrentDialog = dialog;
-            CurrentDialog.State.ValueChanged += state => onDialogOnStateChanged(dialog, state.NewValue);
 
-            dialogContainer.Add(CurrentDialog);
+            Scheduler.Add(() =>
+            {
+                // if any existing dialog is being displayed, dismiss it before showing a new one.
+                lastDialog?.Hide();
+
+                CurrentDialog = dialog;
```
Comment (bdach) [inline osu.Game/Overlays/DialogOverlay.cs:60]: `CurrentDialog` is set to `dialog` twice in this method: once immediately, and once inside the schedule. Is this intended? Scares me a bit, especially what with the `dialog == CurrentDialog` guard at the start of this method that may not work correctly now...

Also if the outer write is the only correct one, then I'd expect to see an "ensure we haven't been replaced already" flow here, as in something along the lines of

```csharp
if (CurrentDialog == dialog)
    // set up value changed callbacks, add to container, show
```

File: osu.Game/Overlays/DialogOverlay.cs [hunk 3]
```diff
@@ -88,16 +94,11 @@ protected override void PopIn()
         protected override void PopOut()
         {
             base.PopOut();
+            this.FadeOut(PopupDialog.EXIT_DURATION, Easing.InSine);
```
Comment (bdach) [inline osu.Game/Overlays/DialogOverlay.cs:97]: Is this reordering intended? Previously the overlay would only fade out if the current dialog was present and visible. May be fine, mostly just wondering.

Comment (peppy) [reply osu.Game/Overlays/DialogOverlay.cs:97]: I actually don't know. The `IsPresent` override on this class makes fading a bit of a mystery.

In testing removing both the fade in and out seems to have zero effect, as that is managed by the individual popups. So I'll just remove both while we're here and see how that plays out.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**